### PR TITLE
[GPU]Stateless LLM: extend CausalMaskAlignment input for gpu::op::SDPA

### PR DESCRIPTION
**Pull request by Kotomi-Du**
### Details:
Currently SDPA with causal LLM only supports upper left attention mask which doesn't work for the LLM decoding scenario where seq_q < seq_kv (new tokens attending to past KV cache). 
This PR covered too parts.
1. Introduce a new input CausalMaskAlignment for gpu::op::SDPA to support LLM stateless scenario. 
2. Add GPU kernel support based on new inputs 
3. Add a GPU-Specific GQA decomposition pass for ONNX GQA operator to utilize the change in `#1` for better optimization

### Tickets:
 - [CVS-187874](https://jira.devtools.intel.com/browse/CVS-187874) ; [CVS-191886](https://jira.devtools.intel.com/browse/CVS-191886)

### AI Assistance:
 - AI assistance used:  yes
 - AI helps create the basic change, and it is reviewed and modified by engineer.

### CODE REVIEW

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp [hunk 1]
```diff
@@ -385,6 +388,25 @@ ov::OutputVector ov::pass::GroupQueryAttentionDecomposition::decompose(
     return {output, present_k, present_v};
 }
 
+std::shared_ptr<ov::Node> ov::pass::GroupQueryAttentionDecomposition::make_sdpa(const ov::Output<ov::Node>& query,
+                                                                                const ov::Output<ov::Node>& key,
+                                                                                const ov::Output<ov::Node>& value,
+                                                                                const ov::Output<ov::Node>& mask,
+                                                                                const ov::Output<ov::Node>& scale,
+                                                                                const ov::Output<ov::Node>& sink,
+                                                                                bool is_causal) {
```
Comment (ankitm3k) [inline src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:404]: @Kotomi-Du  there is a new update in GQA spec for causal support for Bi-Directional GQA below in doc below - https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.GroupQueryAttention
ORT kernel PR - https://github.com/microsoft/onnxruntime/pull/31704

I shall be implementing the same soon

Comment (Kotomi-Du) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:404]: ok, isthere a clarification what kind of mask (upper-left  or bottom-right) they will use if causal = 1? Currently there is an issue for core::SDPA that it treat the mask as upper-left by default. Please check the related discussion in ticket [CVS-191886](https://jira.devtools.intel.com/browse/CVS-191886)

**Comment by Kotomi-Du** [comment]
build_jenkins

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 1 above)
Comment (ankitm3k) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:404]: our current decomposition never sets `is_causal=true` on the public SDPA node for the `causal=1` case — all three SDPA construction sites in `decompose()` pass `is_causal=false` unconditionally, and the causal semantics live entirely in the explicit lower-right-equivalent additive mask built by `make_attention_mask()`. So GPU's alignment default doesn't matter for our path today, unless GPU's own override bypasses that mask for its fast path. Your `make_attention_mask()` override's nullptr-mask short-circuit `(!sliding_window_cache && !external_bias && scale == 0.0f && !is_static_input`) doesn't check causal — once #37653's causal param lands, that condition needs a causal check too, otherwise bidirectional GQA (`causal=0`) would silently take the `is_causal=true` + LOWER_RIGHT fast path meant only for the causal case.

File: src/plugins/intel_gpu/tests/unit/test_cases/sdpa_gpu_test.cpp [hunk 2]
```diff
@@ -282,6 +282,138 @@ TEST_P(sdpa_gpu_test, basic_caching) {
...
+    ASSERT_EQ(ref_ptr.size(), opt_ptr.size());
+    for (size_t i = 0; i < ref_ptr.size(); ++i) {
+        ASSERT_FALSE(std::isnan(static_cast<float>(ref_ptr[i]))) << "NaN in explicit mask output at index " << i;
+        ASSERT_FALSE(std::isnan(static_cast<float>(opt_ptr[i]))) << "NaN in is_causal output at index " << i;
+    }
+
+    const float sim = cosineSimilarity(ref_ptr, opt_ptr);
+    ASSERT_GE(sim, 0.99f) << "explicit mask vs is_causal cosine similarity too low: " << sim;
+}
+
```
Comment (ankitm3k) [inline src/plugins/intel_gpu/tests/unit/test_cases/sdpa_gpu_test.cpp:439]: All 5 new causal-mask test cases below exercise CAUSAL_MASK_LOWER_RIGHT only. Could we add one case that pins the existing UPPER_LEFT (default) path unchanged — e.g. a non-KV-cache seq_q == seq_kv prefill compared against a known-good reference — so a future refactor can't silently regress the default alignment without a test catching it?

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp [hunk 3]
```diff
@@ -332,7 +332,10 @@ ov::OutputVector ov::pass::GroupQueryAttentionDecomposition::decompose(
                                           T,
                                           local_window_size,
                                           external_bias,
-                                          bias_col_offset);
+                                          bias_col_offset,
```
Comment (ankitm3k) [inline src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:343]: The base `make_attention_mask()` now takes sliding_window_cache, scale, is_static_input but never reads them — only the GPU override uses them. nit: but worth marking unnamed/[[maybe_unused]] in the base signature (or a one-line comment noting they're override-only) so it's clear at a glance this isn't dead plumbing.

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 1 above)
Comment (Kotomi-Du) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:404]: yes, will keep in eye on #37653 
what is the user case of bidirectional GQA (causal=0)? why MSFT add such attribute?

**Comment by Kotomi-Du** [comment]
build_jenkins

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 3 above)
Comment (Kotomi-Du) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:343]: updated

File: src/plugins/intel_gpu/tests/unit/test_cases/sdpa_gpu_test.cpp (hunk 2 above)
Comment (Kotomi-Du) [reply src/plugins/intel_gpu/tests/unit/test_cases/sdpa_gpu_test.cpp:439]: updated

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 1 above)
Comment (Kotomi-Du) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:404]: updated based on new change

File: src/plugins/intel_gpu/include/intel_gpu/op/sdpa.hpp [hunk 4]
```diff
@@ -68,6 +73,7 @@ class SDPA : public ov::op::v13::ScaledDotProductAttention {
     std::vector<int64_t> m_order_v;
     std::vector<int64_t> m_order_out;
     ov::element::Type m_output_type;
+    CausalMaskAlignment m_causal_mask_alignment;
```
[the remark this answers is not in the document]
Comment (Kotomi-Du) [reply src/plugins/intel_gpu/include/intel_gpu/op/sdpa.hpp:77]: no change needed, because the default value is set in the initialization function

File: src/plugins/intel_gpu/src/plugin/transformations/op/sdpa.cpp [hunk 5]
```diff
@@ -91,11 +108,15 @@ void SDPA::validate_and_infer_types() {
...
     visitor.on_attribute("order_q", m_order_q);
     visitor.on_attribute("order_k", m_order_k);
     visitor.on_attribute("order_v", m_order_v);
     visitor.on_attribute("order_out", m_order_out);
     visitor.on_attribute("output_type", m_output_type);
+    bool causal_lower_right = m_causal_mask_alignment == CausalMaskAlignment::LOWER_RIGHT;
+    visitor.on_attribute("causal_lower_right", causal_lower_right);
+    m_causal_mask_alignment = causal_lower_right ? CausalMaskAlignment::LOWER_RIGHT : CausalMaskAlignment::UPPER_LEFT;
     return true;
 }
```
[the remark this answers is not in the document]
Comment (Kotomi-Du) [reply src/plugins/intel_gpu/src/plugin/transformations/op/sdpa.cpp:121]: not required, m_is_casual is correctly handled by visitor.on_attribute("is_causal", m_is_causal);

File: src/plugins/intel_gpu/src/plugin/transformations/group_query_attention_decomposition.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    const auto order = op::SDPA::default_order(query.get_partial_shape().rank().get_length());
+    return register_new_node<op::SDPA>(inputs,
+                                       is_causal,
+                                       order,
+                                       order,
+                                       order,
+                                       order,
+                                       ov::element::dynamic,
+                                       op::SDPA::CausalMaskAlignment::LOWER_RIGHT);
```
[the remark this answers is not in the document]
Comment (Kotomi-Du) [reply src/plugins/intel_gpu/src/plugin/transformations/group_query_attention_decomposition.cpp]: updated

File: src/plugins/intel_gpu/src/plugin/transformations/transpose_fusion.cpp [hunk 7]
```diff
@@ -300,7 +304,8 @@ TransposeSDPAMatcher::TransposeSDPAMatcher() {
             inputs.push_back(sdpa->get_input_source_output(4));
         }
 
-        auto sdpa_new = std::make_shared<op::SDPA>(inputs, sdpa->get_causal(), order_q, order_k, order_v, order_output);
+        auto sdpa_new =
+            std::make_shared<op::SDPA>(inputs, sdpa->get_causal(), order_q, order_k, order_v, order_output, ov::element::dynamic, causal_mask_alignment);
```
[the remark this answers is not in the document]
Comment (Kotomi-Du) [reply src/plugins/intel_gpu/src/plugin/transformations/transpose_fusion.cpp]: update

**Review by e-ddykim — COMMENTED** [review]
Attaching a repro test (drop into src/plugins/intel_gpu/tests/unit/transformations/) — 4 pass, 3 fail on this head. The mask elision in the GPU make_attention_mask() override looks too broad.

https://github.com/e-ddykim/openvino/blob/71510e249662e194833cefb692841f5b35342a59/src/plugins/intel_gpu/tests/unit/transformations/gqa_decomposition_sink_repro_test.cpp

**Comment by Kotomi-Du** [comment]
> Attaching a repro test (drop into src/plugins/intel_gpu/tests/unit/transformations/) — 4 pass, 3 fail on this head. The mask elision in the GPU make_attention_mask() override looks too broad.
> 
> https://github.com/e-ddykim/openvino/blob/71510e249662e194833cefb692841f5b35342a59/src/plugins/intel_gpu/tests/unit/transformations/gqa_decomposition_sink_repro_test.cpp

Updated the constraints and related tests to guard.

**Review by e-ddykim — APPROVED** [review]
